lightning/src/ln/channelmanager.rs

```diff
@@ -6180,6 +6180,165 @@ where
+											&self.secp_ctx, b.inbound_blinding_point, &encrypted_tlvs_ss
+										).ok()
+									})
+								});
+
+								let mut full_outgoing_amt_msat = outgoing_amt_msat;
+								let mut full_outgoing_cltv = outgoing_cltv_value;
+
+								let outer_onion_packet = {
+									let path = Path {
+										hops: hops.clone(),
+										blinded_tail: None,
+									};
+									let recipient_onion = RecipientOnionFields::spontaneous_empty();
+									let (mut onion_payloads, htlc_msat, htlc_cltv) = onion_utils::build_onion_payloads(
+										&path,
+										outgoing_amt_msat,
+										&recipient_onion,
+										outgoing_cltv_value,
+										&None,
+										None,
+										None,
+									).unwrap();
+
+									if let Some(last_payload) = onion_payloads.last_mut() {
+										match last_payload {
+											msgs::OutboundOnionPayload::Receive { sender_intended_htlc_amt_msat, cltv_expiry_height, .. } => {
+												*last_payload = match next_blinding_point {
+													None => msgs::OutboundOnionPayload::TrampolineEntrypoint {
+														amt_to_forward: *sender_intended_htlc_amt_msat,
+														outgoing_cltv_value: *cltv_expiry_height,
+														multipath_trampoline_data: None,
+														trampoline_packet: onion_packet.clone(),
+													},
+													Some(blinding_point) => msgs::OutboundOnionPayload::BlindedTrampolineEntrypoint {
+														amt_to_forward: *sender_intended_htlc_amt_msat,
+														outgoing_cltv_value: *cltv_expiry_height,
+														multipath_trampoline_data: None,
+														trampoline_packet: onion_packet.clone(),
+														current_path_key: blinding_point,
+													}
+												};
+											}
+											_ => {
+												unreachable!("Last element must always initially be of type Receive.");
+											}
+										}
+									}
+
+									let outer_session_priv = SecretKey::from_slice(&self.entropy_source.get_secure_random_bytes()).unwrap();
+									let onion_keys = onion_utils::construct_onion_keys(&self.secp_ctx, &path, &outer_session_priv).map_err(|_| {
+										APIError::InvalidRoute { err: "Pubkey along hop was maliciously selected".to_owned() }
+									}).unwrap();
+									let outer_onion_prng_seed = self.entropy_source.get_secure_random_bytes();
+									let onion_packet = onion_utils::construct_onion_packet(onion_payloads, onion_keys, outer_onion_prng_seed, &payment_hash).unwrap();
+
+									full_outgoing_amt_msat = htlc_msat;
+									full_outgoing_cltv = htlc_cltv;
+
+									onion_packet
+								};
+
+								// Forward the HTLC over the most appropriate channel with the corresponding peer,
+								// applying non-strict forwarding.
+								// The channel with the least amount of outbound liquidity will be used to maximize the
+								// probability of being able to successfully forward a subsequent HTLC.
+								let maybe_optimal_channel = peer_state.channel_by_id.values_mut()
```

Comment:
I'd not expect this code to be repeated, because it also applies to non-trampoline forwards? I do realize this is draft.